### ISSUE TITLE
kobo enhancement: handle usb_storage from within koreader

### ISSUE THREAD

**Issue by pazos** [koreader/koreader#2597]
KOreader works well as the main software in a kobo ereader. The only basic feature I miss during my readings is to be able to plug the ereader into a pc and transfer some files.

kobo rootfs has some udev rules that trigger some scripts under some circunstances (sd, usb, ac, plug). All those scripts doesn't mean nothing from outside nickel, because they write strings to  /tmp/nickel_hardware_status (from memory)

I have no idea how to implement this :1234:

**Comment by houqp** [comment]
At this point, this is actually pretty straight forward to implement because I already did the USB hotplug detection part, which is needed for one of the suspend issue fix. See:

https://github.com/koreader/koreader-base/blob/master/input/input-kobo.h#L43
https://github.com/koreader/koreader/blob/master/frontend/ui/uimanager.lua#L101

Basically we can just show up a UI to confirm usb_storage mount in "Charging" event handler.

**Comment by Hzj-jie** [comment]
Out of topic, but this can be achieved by using sftp & keep alive plugin.

**Comment by pazos** [comment]
@Hzj-jie: yes!, or smb, or netatalk. I tried all of them and they are great alternatives, but just that.

I had a paperwhite and I was very happy with koreader, because the interaction between system software and extensions. The whole intercommunication system was great and I had access to system features (as wifi, usb mass storage or suspend capabilities) from within extensions.

In kobos: No interaction with nickel at all. In some cases we kill kobo software before running and try to run it again at exit, but thats all. Every basic feature has been implemented in koreader: suspend, wifi, frontlight. Just one remains unimplemented: usb mass storage.

> Basically we can just show up a UI to confirm usb_storage mount in "Charging" event handler.

@houqp: I was happy with a menu entry because I thought it was easier to implement, and less intrusive also, but the ui sounds great. :+1: 

the stock os scripts for usb mass storage can be repurposed for this with little effort, but we need to change the mountpoint from /mnt/onboard to /mnt/onboard/somefolder to avoid umounting koreader's parent folder or install koreader in our rootfs (/mnt/koreader or /opt/koreader, for example).

**Comment by houqp** [comment]
Oh yeah, a menu entry would do it too, just in a less automated fashion, but still a huge improvement to what we have right now. I still exit to KSM and enable usb storage from there every time, which is annoying.

If you can help prepare the mount script, adding a menu to call that script should be a trivia patch.

**Comment by poire-z** [comment]
> or install koreader in our rootfs (/mnt/koreader or /opt/koreader, for example).

I hope we can avoid that, that would make it less easy to hack/test on the device, we would need some kind of copy from/to usb from/to rootfs stuff/tricks, like KSM 8 did/needs (I'm still using KSM 7 just because of that).

**Comment by houqp** [comment]
Let's try mounting it to another path first, I would also prefer KOReader stay in an easy to remove/update user accessible directory too.

**Comment by pazos** [comment]
I've been busy at work these days, but I did a few tests sharing the partition via ums without unmounting it and it is safe. Very safe for koreader indeed.

I got a few data corruption during heavy i/o tasks from both sides at the same time, like:

from computer: 
`while true; do dd if /dev/random of=/path/of/usb_share/test.file bs=256m count=10; done`

from the ereader
`count=1`
 `while true; do sleep "$count"; rm -rf /mnt/onboard/test.file; count=$(count)+1;  done`

The few brokes I got were: `I/O error` and recoverable with e2fsck, which runs by default on kobo firmware.

I think it is safe to implement a button to enable usb storage without unmounting any partition. In any case it is a must to disable wifi, suspend and auto-suspend/shutdown, the frontlight, and any other time-triggered events happening from koreader during the plug (in udev from ADD to REMOVE)

And the workflow: 
with usb storage disabled = nothing // no kernel module loaded
with usb storage enabled = share the mountpoint /mnt/onboard via ums kernel driver

if usb storage enabled then wait for a "usb" event to enter into a safe mode -read above-. then wait for the unplug event to return to normal operation.

**Comment by KenMaltby** [comment]
Will the external SD be included as well? Mountpoint /mnt/sd.

**Comment by houqp** [comment]
That's good new :)

> I think it is safe to implement a button to enable usb storage without unmounting any partition. In any case it is a must to disable wifi, suspend and auto-suspend/shutdown, the frontlight, and any other time-triggered events happening from koreader during the plug (in udev from ADD to REMOVE)

We can disable wifi, suspend, auto-suspend/shutdown and frontlight. But I am not sure how feasible it is to prevent all time-triggered events from happening. For example, when a user touches the screen, the kernel will still deliver a touch event right? I assume this measure is to prevent KOReader from writing into /mnt/onboard?

@KenMaltby I haven't used SD before, we can look into this once the internal storage mount is working ;)

**Comment by KenMaltby** [comment]
When you "connect", via the usb cable, in both Nickel and KSM Windows sees the device as a "mass storage device" and creates drives for both the internal and external flash storage devices.  KSM offers several options for that, to include a usbnet connection.  I assume all we are looking for here is the basic mass storage connection like nickel and KSM provide.

**Comment by lgeek** [comment]
How difficult would it be to get Koreader in a state where it's guaranteed not to write to any files? If that's practical, then it would be possible to remount the data partition RO on the device, expose it over USB, then on eject flush the disk cache and remount the partition RW. Otherwise, it might be easier to just close koreader and use a separate application / script to restart koreader on eject.

**Comment by pazos** [comment]
> How difficult would it be to get Koreader in a state where it's guaranteed not to write to any files?

@lgeek: good question. During my tests autosuspend was the feature causing trouble and leads the device to an unknown state.

> I assume this measure is to prevent KOReader from writing into /mnt/onboard?

@houqp: yes, the point is avoid /mnt/onboard corruption. This is pretty easy to achieve if we can unmount and fsck /mnt/onboard (even having a fallback method for corrupted data), but having koreader installed under this path makes the thing a little bit anoying

But disabling frontlight and wifi (to save battery cycles), and suspend/autosuspend (for the above reasons) should be enough.
Another HUGE break condition is to leave koreader during usb transfers. Nickel doesn't expect an environment like that, it will try to umount and fsck, will find a bad block and force an "exit account" soft-factory-reset.

So, in short:
 1. a button: "Enable USB Mass Storage"
 2. when the button is checked wait for udev events (plug?)
 3. when a new event arrives put the device in "usb_transfer_mode:nowifi,no(auto)suspend,nolight,graphical_feedback
4. finally load and setup the kernel module to expose /mnt/onboard (and /mnt/sd if present) via usb.

4 is fairly easy to do. But 2 and 3 seems to much for me :1st_place_medal: 

As a side note, once usb_mass_storage is implemented in the UI, making another button to enable usbnet in the same fashion will be super-easy

**Comment by pazos** [comment]
another bunch of tests in my free day :)

Until now I've tested under an optimal environment (ie: a good cable attached between pc and device and don't touch this particular connection). Just needed a momentary usb disconnection to break the device.

So it isn't safe to handle usb storage mountpoints that we can't check. This makes /mnt/onboard non exportable via usb in most installations, but exportable in custom installations where koreader is installed  on another mountpoint

KSM also unmount /mnt/onboard before exporting and fsck /dev/mmcblk0p3 after exporting.

**Comment by pazos** [comment]
Just an update on this issue:

@baskerville 's [plato](https://github.com/baskerville/plato) has support for USBMS and his scripts work on koreader too :smile: As @houqp suggested we can call [usb-enable.sh](https://github.com/baskerville/plato/blob/master/scripts/usb-enable.sh) on "Charging" event and [usb-disable.sh](https://github.com/baskerville/plato/blob/master/scripts/usb-disable.sh) on "Not Charging" in [frontend/ui/uimanager.lua](https://github.com/koreader/koreader/blob/master/frontend/ui/uimanager.lua#L130-L142).

Not sure about the UI. Maybe

```
                if G_reader_settings:isTrue("usb_mass_storage") then
                    local ConfirmBox = require("ui/widget/confirmbox")
                    UIManager:show(ConfirmBox:new{
                        text = _("Share storage via USB?\n"),
                        ok_text = _("Share"),
                        ok_callback = function()
                            -- stop hw buttons
                            -- stop frontlight
                            -- stop wifi
                            -- stop suspend ( & background runner ?? )
                            -- quit current book (go to FM app)
                            os.execute(./usb-enable.sh)
                        end,
                     })
                end
```

**Comment by Frenzie** [comment]
How long does it block?

**Comment by pazos** [comment]
> How long does it block?

Not sure if I understand. The script "usb-enable.sh" just export storage over usb. It doesn't block by itself. We need to make sure that other stuff (like hw buttons & autosuspend) are unscheduled before launching that script.

Sadly our USB Hotplug detection doesn't catch differences between AC and USB_HOST so it will display the confirmbox always (even when connected to the wall). Not sure how plato is doing this because it works fine there (the box is shown only when attached to a usb host)

Need to tinker a bit with this code to see if it can catch differences:

```
// Device information for ntx platform. Tested on Kobo Aura HD & Bq Cervantes 4

#include <stdio.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>

// power
#define	AC_IN				    118
#define CM_USB_Plug_IN			108
#define	CM_USB_AC_STATUS		122
#define	CM_GET_BATTERY_STATUS 	206
#define	CM_RTC_WAKEUP_FLAG      123

int main(int argc,char **argv)
{
    unsigned long retval,pointer,value;

    int ntx_io=open("/dev/ntx_io",O_RDWR);
    
    if (!ntx_io)
    {
        fprintf(stderr, "Error: can't read from /dev/ntx_io, exiting...");
        return 1;
    } 

    // AC_IN - USB is inserted?
    retval=ioctl(ntx_io, AC_IN, &pointer);
    printf("ioctl(ntx_io,AC_IN)                  = %d; returned %ld\n",retval,pointer);

    // CM_USB_Plug_IN - USB is inserted?
    retval=ioctl(ntx_io, CM_USB_Plug_IN, &pointer);
    printf("ioctl(ntx_io,CM_USB_Plug_IN)         = %d; returned %ld\n",retval,pointer);

    // CM_USB_AC_STATUS - (always 0)?
    retval=ioctl(ntx_io, CM_USB_AC_STATUS, &pointer);
    printf("ioctl(ntx_io,CM_USB_AC_STATUS)       = %d; returned %ld\n",retval,pointer);

    // CM_GET_BATTERY_STATUS - battery charge?
    retval=ioctl(ntx_io, CM_GET_BATTERY_STATUS, &pointer);
    printf("ioctl(ntx_io,CM_GET_BATTERY_STATUS)  = %d; returned %ld\n",retval,pointer);

    // CM_RTC_WAKEUP_FLAG - Woken by RTC?
    retval=ioctl(ntx_io, CM_RTC_WAKEUP_FLAG, &pointer);
    printf("ioctl(ntx_io,CM_RTC_WAKEUP_FLAG)     = %d; returned %ld\n",retval,pointer);

    close(ntx_io);
    return 0;
}
```

**Comment by baskerville** [comment]
> Sadly our USB Hotplug detection doesn't catch differences between AC and USB_HOST so it will display the confirmbox always (even when connected to the wall). Not sure how plato is doing this because it works fine there (the box is shown only when attached to a usb host)

I read the plug/unplug events from `/tmp/nickel-hardware-status`.

The corresponding messages are `usb plug {add,remove}`.

**Comment by Frenzie** [comment]
@pazos 
> Not sure if I understand. The script "usb-enable.sh" just export storage over usb. It doesn't block by itself. We need to make sure that other stuff (like hw buttons & autosuspend) are unscheduled before launching that script.

If it blocks it would probably be a good idea to do something about that. If it doesn't block then the answer is simply that it blocks for <1s and your suggested code could be fine as is.

**Comment by pazos** [comment]
> I read the plug/unplug events from /tmp/nickel-hardware-status. The corresponding messages are usb plug {add,remove}.

Oh great, thanks!. I found this on mobileread: https://www.mobileread.com/forums/showpost.php?p=3338563&postcount=33. That means that we can skip the /tmp/nickel-hardware-status and generate the fake events from https://github.com/koreader/koreader-base/blob/master/input/input-kobo.h. Right now we only use that file to trigger a suspend after an usb plug/unplug event so we don't differenciate between usb_host and usb_plug.

@Frenzie should I add this as part of [koreader-base/pull/745](https://github.com/koreader/koreader-base/pull/745) or add a new PR instead?

**Comment by NiLuJe** [comment]
New PR, please ;)

**Comment by Frenzie** [comment]
Yeah, better separately. :-)

**Comment by RobinFrcd** [comment]
Hi,
Is there a way to transfer file from PC to Koreader once the system is started ? Thanks!

**Comment by Frenzie** [comment]
You can use the built-in SSH server to connect using an FTP program such as Filezilla.

**Comment by KenMaltby** [comment]
There are also wireless methods, for ebook files, the easiest being those using calibre,

**Comment by louisabraham** [comment]
Hi, what's the current status of this issue? I have a kobo ereader and would like to activate USB from koreader, or at least go back to KSM without having to restart.

**Comment by NiLuJe** [comment]
Probably never going to happen. Much, much, much too brittle without a custom setup (i.e., KOReader living somewhere else than onboard, which is unsupported, and probably impossible on an unmodified device because of the limited free space available).

**Comment by NiLuJe** [comment]
Going back to KSM on exit technically shouldn't require a restart, though. That said, not familiar with that, so, I'll leave the details to people who actually know how KSM works.

**Comment by Frenzie** [comment]
My / has 99.3M available, which is rather tight but perhaps not impossible. :-P (NB I strongly recommend against trying it.)

You can exit KOReader to KSM, toggle USB on, do your thing, toggle USB off, and restart KOReader. It's a few mildly annoying extra steps which is part of why I use the SSH server over wifi instead, although mainly it's because I don't want to go digging for cables.

**Comment by louisabraham** [comment]
Thank you so much for the quick answer!

I saw a couple of people talking about ssh but couldn't find any instruction on the wiki. Maybe https://github.com/koreader/koreader/wiki/SSH could be updated? In particular, I'm lacking the login and password (root as login was denied).

How do you exit KOReader? Sorry if this is off-topic but I have used koreader for 2 years without knowing.

Thank you again for your help!

**Comment by NiLuJe** [comment]
There's an option to disable password checks in plugin's setting. If that's enabled, you can simply login as root with an empty password field.

**Comment by Frenzie** [comment]
You have to select "login without password" in the SSH menu. Then you can transfer files using most FTP clients with a configuration like this:

![Screenshot_2019-08-12_15-11-20](https://user-images.githubusercontent.com/202757/62867437-8a126380-bd13-11e9-8d08-4ebd5f25dcf8.png)

The IP is device-specific, the port defaults to 2222.

Or in the addressbar of the FTP client, `sftp://root@192.168.1.213:2222`.

The process as described in the wiki is much more niche, although Lsyncd should work fairly conveniently once set up.

**Comment by louisabraham** [comment]
Oh, thanks, I didn't know about port 2222! IMO it would be very useful to put the content of your messages in the wiki.

How do you exit koreader?

**Comment by pazos** [comment]
> Probably never going to happen. Much, much, much too brittle without a custom setup (i.e., KOReader living somewhere else than onboard, which is unsupported, and probably impossible on an unmodified device because of the limited free space available).

@NiLuJe: KOReader can still live on /mnt/onboard. Just need to install a few scripts  (usbms-enable.sh, usbms-disable.sh and a quick executable called whatever) on rootfs and invoke whatever from KOReader.

On cervantes we rely on a button press to unmount the usbms thingy because AFAICT udev doesn't alert us about usb hotplug events, but in the case on kobos I think you can watch for the nickel_hardware in /tmp and react accordly.

**Comment by Frenzie** [comment]
@louisabraham Well, it's in the "start ssh server' popup but I do think it's (too) easy to miss. ;-)

I'd put it on the wiki but I'm not sure which FTP client to recommend. I understand that on Windows FileZilla has malware issues with the installer.

@pazos There's already some preliminary support for that:

https://github.com/koreader/koreader/blob/53b77e3ae4d3fec39628cfb286f000908cdeaa8f/frontend/device/kobo/device.lua#L320-L322

https://github.com/koreader/koreader/blob/53b77e3ae4d3fec39628cfb286f000908cdeaa8f/frontend/device/generic/device.lua#L259-L263

**Comment by NiLuJe** [comment]
I still feel like there's way too much potential for disaster w/ a "live" onboard :/.

**Comment by louisabraham** [comment]
@Frenzie Actually, I don't even have the ssh plugin. I probably need to upgrade koreader! 

FileZilla is perfect on my macOS. For windows [WinSCP](https://winscp.net/eng/download.php) might be a good alternative.


How can I exit koreader without restarting?

**Comment by NiLuJe** [comment]
> Exit is in the hamburger menu (final tab on the right in the top menu, whatever the context).

Or not, if you're running a truly ancient version, which seems to be the case if you don't have the SSH plugin.

**Comment by Frenzie** [comment]
NB You don't find the SSH plugin under plugins/tools (anymore) but under the network menu.

![Screenshot_2019-08-12_15-42-27](https://user-images.githubusercontent.com/202757/62869299-de1f4700-bd17-11e9-946c-1940853e1069.png)

**Comment by Frenzie** [comment]
### Exit
![Screenshot_2019-08-12_15-44-52](https://user-images.githubusercontent.com/202757/62869444-29d1f080-bd18-11e9-95d7-ea49e05e3eb2.png)

**Comment by AlanSP1** [comment]
@louisabraham 

> Oh, thanks, I didn't know about port 2222! IMO it would be very useful to put the content of your messages in the wiki.

OK, I added additional info in wiki: https://github.com/koreader/koreader/wiki/SSH

Do you think there's more info that needs to be added? Is it clear enough and if not, what's missing, how it can be clear enough?

**Comment by Frenzie** [comment]
I edited it to say SFTP instead of FTP.

Btw, in Dolphin and I think Nautilus too you can just type `sftp://root@192.168.1.213:2222` like I said above. In Windows I have a license for Directory Opus which can do that too. Windows Explorer can do it for FTP, but not for SFTP unless that changed in recent years. 

A dedicated client like FileZilla has nice features like multiple simultaneous connections and pausing transfers, but on your local network that kind of thing seldom matters.

![Screenshot_20190812_235153](https://user-images.githubusercontent.com/202757/62901256-97a00b80-bd5c-11e9-847f-198b5370e4d8.png)
![Screenshot_20190812_235428](https://user-images.githubusercontent.com/202757/62901258-9838a200-bd5c-11e9-83b7-712f411f3381.png)

**Comment by potash** [comment]
I understand from the above that it is difficult/impossible to add USB support within koreader. I wonder whether you can streamline the process of USB transfer outside of koreader. It currently requires 9 clicks (Menu > three stripes > Exit > Exit > usb > enable > disable > return > start). Is there any reason it couldn't be done in 4? e.g. (Menu > Three stripes > Exit and Enable USB > Disable USB and Start Koreader).

**Comment by pazos** [comment]
> I understand from the above that it is difficult/impossible to add USB support within koreader.

@potash: no impossible but you need to write a program that handles the mount/unmount and that lives outside /mnt/onboard, so it can be called by KOReader when the user ticks "enable usb".

We do something like this on cervantes because there are no other 3rd party apps there. A few bites to call a 3rd party app to do the job are already available in https://github.com/koreader/koreader/blob/master/frontend/ui/elements/mass_storage.lua and implemented for the cervantes platform in https://github.com/koreader/koreader/blob/master/platform/cervantes/koreader.sh#L104-L110.

>  I wonder whether you can streamline the process of USB transfer outside of koreader. It currently requires 9 clicks (Menu > three stripes > Exit > Exit > usb > enable > disable > return > start). Is there any reason it couldn't be done in 4? e.g. (Menu > Three stripes > Exit and Enable USB > Disable USB and Start Koreader).

I think you're talking about KSM. That's a program outside koreader and nothing we can do *inside* koreader will alter the way it works.

**Comment by AlanSP1** [comment]
KSM is great addition and I use it and love it. But, problem is that tshering stopped (hopefully only for now) maintaining it. :(

**Comment by pazos** [comment]
Closing this issue as won't fix. Feel free to submit PR based on a -optional- 3rd party tool. You can use https://github.com/koreader/koreader/commit/301aa580cc2e37b4ff1eaea023acb1194304d778#diff-23f70819489ec049a1e1e600dcdf57d4 as an example.
